File: JAVA/024_Lexicographic_Permutations.java


import java.util.ArrayList;
import java.util.List;

class TestMain 
{

	
	public static void main(String[] args) 
	{
		List<Integer> inputChars = new ArrayList<Integer>();
		for(int counter=0;counter<10;counter++)
			inputChars.add(counter);
		
		int rank = 1000000;
		
		System.out.println(findPermutation(inputChars, rank));
		
	}
	
	private static String findPermutation(List<Integer> inputChars, int rank) 
	{
		String ranks=findRank(inputChars, rank);
		int counter,individualRank=0;
		StringBuilder finalOutput = new StringBuilder();
		for(counter=0;counter<ranks.length();counter++)
		{
			individualRank=ranks.charAt(counter)-48;
			finalOutput.append(inputChars.remove(individualRank-1));
			
		}
		return finalOutput.toString();
	}

	private static String findRank(List<Integer> inputChars,int rank)
	{
		StringBuilder ranks=new StringBuilder();
		int counter,individualRank=0,x;
		int factorial=1;
		List<Integer> factorials = new ArrayList<Integer>();
		factorials.add(1);
		for(counter=1;counter<=inputChars.size();counter++)
		{
			factorial*=counter;
			factorials.add(factorial);
		}
		
		if(rank>factorials.get(factorials.size()-1))
			return "error";
		
		for(counter=factorials.size()-2;counter>=0;counter--)
		{
			x = rank%factorials.get(counter);
			if(x>0)
			{
				individualRank=(rank/factorials.get(counter))+1;
			}
			else
			{
				individualRank=(rank/factorials.get(counter));
				ranks.append(individualRank);
				for(int inner=--counter;inner>=0;inner--)
				{
					ranks.append(inner+1);
				}
				break;
			}
			ranks.append(individualRank);
			rank=x;
			
		}
		return ranks.toString();
		
	}
}

File: CPP/024_Lexicographic_Permutations.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<char> findPermutation(vector<int> inputChars, int rank);
vector<char> findRank(vector<int> inputChars,int rank);

        int main() 
        {
                vector<char>::size_type i;
                vector<int> inputChars;
                for(int counter=0;counter<10;counter++)
                        inputChars.push_back(counter);
                
                int rank = 1000000;
                
                vector<char> answer = findPermutation(inputChars, rank);
                for(i=0;i<answer.size();i++)
                    cout << answer[i] << "\t";
                
        }
        
        vector<char> findPermutation(vector<int> inputChars, int rank) 
        {
                vector<char> ranks=findRank(inputChars, rank);
                int counter,individualRank=0;
                vector<char> finalOutput;
                for(counter=0;counter<ranks.size();counter++)
                {
                        individualRank=ranks[counter]-48;
                        finalOutput.push_back(inputChars[individualRank-1]+48);
                        inputChars.erase(inputChars.begin()+individualRank-1);
                }
                return finalOutput;
        }
 
        vector<char> findRank(vector<int> inputChars,int rank)
        {
                 vector<char> ranks,error;
                 error.push_back('e');
                 error.push_back('r');
                 error.push_back('r');
                 error.push_back('o');
                 error.push_back('r');
                int counter,individualRank=0,x;
                int factorial=1;
                vector<int> factorials;
                factorials.push_back(1);
                for(counter=1;counter<=inputChars.size();counter++)
                {
                        factorial*=counter;
                        factorials.push_back(factorial);
                }
                
                if(rank>factorials[factorials.size()-1])
                        return error;
                
                for(counter=factorials.size()-2;counter>=0;counter--)
                {
                        x = rank%factorials[counter];
                        if(x>0)
                        {
                                individualRank=(rank/factorials[counter])+1;
                        }
                        else
                        {
                                individualRank=(rank/factorials[counter]);
                                ranks.push_back(individualRank+48);
                                for(int inner=--counter;inner>=0;inner--)
                                {
                                        ranks.push_back(inner+49);
                                }
                                break;
                        }
                        ranks.push_back(individualRank+48);
                        rank=x;
                        
                }
                return ranks;
                
        }

 